pinot-plugins/pinot-file-system/pinot-adls/src/main/java/org/apache/pinot/plugin/filesystem/AzureGen2PinotFS.java

```diff
@@ -0,0 +1,447 @@
+
+          // Compute the destination path for the current path.
+          String relativeSrcPath = srcPath.relativize(Paths.get(convertAzureStylePathToUriStylePath(path))).toString();
+          String newDstPath = Paths.get(dstUri.getPath(), relativeSrcPath).toString();
+          URI newDst = new URI(dstUri.getScheme(), dstUri.getHost(), newDstPath, null);
+
+          if (isDirectory(currentSrc)) {
+            // If src is directory, create one.
+            if (!mkdir(newDst)) {
+              return false;
+            }
+          } else {
+            // If src is a file, we need to copy.
+            copySucceeded |= copySrcToDst(currentSrc, newDst);
+          }
+        }
+        return copySucceeded;
+      } catch (DataLakeStorageException | URISyntaxException e) {
+        throw new IOException(e);
+      }
+    }
+  }
+
+  @Override
+  public boolean exists(URI fileUri) throws IOException {
+    try {
+      _fileSystemClient.getDirectoryClient(convertUriToAzureStylePath(fileUri)).getProperties();
+      return true;
+    } catch (DataLakeStorageException e) {
+      if (e.getStatusCode() == NOT_FOUND_STATUS_CODE) {
+        return false;
+      }
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public long length(URI fileUri) throws IOException {
+    try {
+      PathProperties pathProperties =
+          _fileSystemClient.getDirectoryClient(convertUriToAzureStylePath(fileUri)).getProperties();
+      return pathProperties.getFileSize();
+    } catch (DataLakeStorageException e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public String[] listFiles(URI fileUri, boolean recursive) throws IOException {
+    LOGGER.info("listFiles is called with fileUri='{}', recursive='{}'", fileUri, recursive);
+    try {
+      ListPathsOptions options =
+          new ListPathsOptions().setPath(convertUriToAzureStylePath(fileUri)).setRecursive(recursive);
+      PagedIterable<PathItem> iter = _fileSystemClient.listPaths(options, null);
+      return iter.stream().map(PathItem::getName).toArray(String[]::new);
+    } catch (DataLakeStorageException e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public void copyToLocalFile(URI srcUri, File dstFile) throws Exception {
+    LOGGER.info("copyToLocalFile is called with srcUri='{}', dstFile='{}'", srcUri, dstFile);
+    if (dstFile.exists()) {
+      if (dstFile.isDirectory()) {
+        FileUtils.deleteDirectory(dstFile);
+      } else {
+        FileUtils.deleteQuietly(dstFile);
+      }
+    }
+
+    MessageDigest md5File = MessageDigest.getInstance("MD5");
+    int bytesRead;
+    byte[] buffer = new byte[BUFFER_SIZE];
+    try (InputStream inputStream = open(srcUri)) {
+      try (OutputStream outputStream = new FileOutputStream(dstFile)) {
+        while ((bytesRead = inputStream.read(buffer)) != -1) {
+          outputStream.write(buffer, 0, bytesRead);
+          md5File.update(buffer, 0, bytesRead);
```

Comment:
Would be good to have perf check here, given that we don't do such checks for other PinotFS implementations.

Reply:
will check md5 hash compute time but i think that this should be very quick